retrofit2/src/main/java/com/linecorp/armeria/client/http/retrofit2/ArmeriaCallFactory.java

```diff
@@ -129,80 +135,55 @@ public Request request() {
             return request;
         }
 
-        private synchronized CompletableFuture<AggregatedHttpMessage> createRequest() {
-            if (executed) {
-                throw new IllegalStateException("Already Executed");
+        private synchronized void createRequest() {
+            if (httpResponse != null) {
+                throw new IllegalStateException("executed already");
             }
-            final CompletableFuture<AggregatedHttpMessage> future = doCall(callFactory.httpClient, request)
-                    .aggregate();
-            executed = true;
-            return future;
+            executionStateUpdater.compareAndSet(this, ExecutionState.IDLE, ExecutionState.RUNNING);
+            httpResponse = doCall(callFactory.httpClient, request);
         }
 
         @Override
         public Response execute() throws IOException {
-            future = createRequest();
+            CompletableCallback completableCallback = new CompletableCallback();
+            enqueue(completableCallback);
             try {
-                return convertResponse(future.get());
-            } catch (InterruptedException e) {
+                return completableCallback.join();
+            } catch (CancellationException e) {
                 throw new IOException(e);
-            } catch (ExecutionException e) {
+            } catch (CompletionException e) {
                 throw new IOException(e.getCause());
             }
         }
 
         @Override
-        public void enqueue(Callback responseCallback) {
-            future = createRequest();
-            future.whenComplete((response, throwable) -> {
-                if (throwable != null) {
-                    responseCallback.onFailure(this, new IOException(throwable.getMessage(), throwable));
-                } else {
-                    if (!canceled) {
-                        try {
-                            responseCallback.onResponse(this, convertResponse(response));
-                        } catch (IOException e) {
-                            responseCallback.onFailure(this, e);
-                        }
-                    } else {
-                        responseCallback.onFailure(this, new IOException("Canceled"));
-                    }
-                }
-            });
+        public void enqueue(Callback callback) {
+            createRequest();
+            httpResponse.subscribe(new ArmeriaCallSubscriber(this, callback, request));
         }
 
         @Override
         public void cancel() {
-            canceled = true;
-            if (future != null) {
-                future.cancel(true);
-            }
+            executionStateUpdater.set(this, ExecutionState.CANCELED);
         }
 
         @Override
         public synchronized boolean isExecuted() {
```

Comment:
Given that `httpResponse` is volatile, we could remove the `synchronized` modifier here.